docs/DesignDocs/shared_memory.md

```diff
@@ -0,0 +1,192 @@
+The Open Enclave shared memory model
+====================================
+
+This document proposes a *shared memory model* that will improve
+interoperability across different secure hardware platforms, initially
+including SGX and TrustZone.
+
+Motivation
+----------
+
+Ideally, applications developed with the Open Enclave API should run on all
+supported platforms. But unfortunately, the platforms themselves may have
+incompatible shared memory models. For example, SGX implicit shares all
+untrusted memory with the enclave, whereas TrustZone shares untrusted memory
+explicitly through programmatic measures (by default, no untrusted memory
+is shared).
+
+This problem is partly addressed by the PR entitled [Enforcing full
+serialization in EDL](https://github.com/openenclave/openenclave/pull/2176),
+which proposes extending **edger8r** to print warnings for EDL specifications
+with parameters that cannot be fully serialized. Such specifications produce
+applications that implicitly share untrusted memory with the enclave. Aside
+from introducing potential security flaws, these applications are not portable
+to TrustZone and any future platforms that employ a similar model. By forcing
+full serialization, the need for shared memory can be mostly eliminated. But
+some applications need to pass untrusted shared memory to enclaves. Examples
+include:
+
+- Ring buffers
+- Context-switchless calls
+- Efficient transfer of large memory objects
```

Comment:
Shared memory increases risk of [TOCTOU](https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use) vulnerabilities. While the enclave is processing the shared the object, a malicious host could change the contents of the shared memory object, potentially causing security vulnerabilities. A common strategy to prevent TOCTOU issues is to copy the object to enclave memory, operate on it and then to copy it (results) out back to host memory.

It is not clear how shared-memory would help with efficient transfer of large memory objects since for secure coding, these objects must be copied to enclave memory.

Reply:
Some use cases are read-once use cases (no check) and so are not susceptible to TOCTOU.   Read-once use cases can be done without copying.   I agree it would be helpful to elaborate on this point with the issues you point out.